### ISSUE TITLE
The tuple built-in is not supported in nopython mode

### ISSUE THREAD

**Issue by krrk** [numba/numba#2771]
The `tuple([iterable])` python built-in function doesn't work in nopython mode. Is there any plan to support this in nopython mode?

This would be very helpful for some functions like `numpy.ndindex(*shape)` which in numba must be called with a tuple. In addition in nopython mode generating a tuple by tuple addition (concatenation) in a for loop does not work.

Minimal function for use of tuple:
```python
@jit(nopython=True)
def g():
    tuple()
```
Resulting error:
```python
Untyped global name 'tuple': cannot determine Numba type of <class 'type'>
```
Trying to generate a tuple using a for loop:
```python
@jit(nopython=True)
def foo():
    t = ()
    for i in range(-10, 11):
        t += (i,)
    return t
```
Resulting error:
```python
cannot unify () and (int64 x 1) for 't'
```

Also after reading through the documentation (section 2.6) I had expected that this would work. Here are the relevant parts of that documentation:

> 2.6. Supported Python features¶
> 
> Apart from the Language part below, which applies to both object mode and nopython mode, this page only lists the features supported in nopython mode.

> 2.6.2.3. tuple
> 
> The following operations are supported:
> 
>     tuple construction
>     tuple unpacking
>     comparison between tuples
>     iteration and indexing over homogenous tuples
>     addition (concatenation) between tuples
>     slicing tuples with a constant slice
>

**Comment by ehsantn** [comment]
I wish these cases were easier but `tuple([iterable])` is tricky since Numba needs to know the tuple length and data types for compilation. This is not possible in general.

However, we should think about cases that the tuple type can be inferred statically.

**Comment by Rik-de-Kort** [comment]
> is tricky since Numba needs to know the tuple length and data types for compilation. This is not possible in general.

This strikes me as a rather odd comment. List() is supported fine, so why isn't tuple()? List types are inferred fine, then, and their types shouldn't be all that different (though I'm new and I don't have access to the exact type specification).

More to the point at OP: 
* I'm not a computer I can test on right now, but does it have to be a tuple? List unpacking is a thing.
* Your second example runs into a typing error because the numba type for a zero-length tuple is different than the numba type for a tuple of length one. So if you use a new variable, the tuple addition will work fine. Honestly, I'm not sure why the tuple type should have fixed lengths, as long as it's immutable (for which there is a flag set for lists...).

**Comment by krrk** [comment]
> This strikes me as a rather odd comment. List() is supported fine, so why isn't tuple()?

Your statement surprised me because I had tested generating an empty list by replacing tuple with list in my first example which fails with a TypingError. It seems numba doesn't know what type list to generate and in reality there is no sane default option. If you do pass a list or tuple (with all elements of compatible types? It seems int and float work together, but not string and int or string and float.) to the list built-in then yes it does work; however, this doesn't work for tuple.

> but does it have to be a tuple? List unpacking is a thing.

For nopython mode the current answer is yes. A lot of the numpy functions require a shape parameter (e.g. numpy.zeros, numpy.ones, numpy.reshape) for which currently a tuple must be passed. Since there is no way to generate the necessary shape tuple in nopython mode curretly it has to be passed into the jitted function as a parameter. Here is an example:

```python
@jit(nopython=True)
def g():
    np.zeros([1,2,3])
```
give the following error
```python
numba.errors.TypingError: Failed at nopython (nopython frontend)
Invalid usage of Function(<built-in function zeros>) with parameters (list(int64))
```

This is perhaps not a big deal, but as a new user of numba and a fundamentally python user (I'm not used to using other languages) it is quite jarring to run into many situations where the pythonic way of programming has failed when attempting to use numba in nopython mode to gain the best speedup.

> Your second example runs into a typing error because the numba type for a zero-length tuple is different than the numba type for a tuple of length one. So if you use a new variable, the tuple addition will work fine. Honestly, I'm not sure why the tuple type should have fixed lengths, as long as it's immutable (for which there is a flag set for lists...).

My second example was an attempt to get around the lack of the tuple built-in. I don't see a way of generating an arbitrary length tuple by using another variable because it seems that numba tries to infer the type and length of the declared tuple and cannot dynamically change it. So for example if I use another variable:
```python
@jit(nopython=True)
def foo():
    t = ()
    for i in range(10):
        s = t + (i,)
        t = s
    return t
```
it'll still fail with a TypingError.

**Comment by krrk** [comment]
@ehsantn I agree with @Rik-de-Kort about the type inferencing. It seems that numba can determine the type and length of lists in nopython mode and if that is known then for at least those cases it should be easy to create a tuple from a list.

For example in the function below numba can create the list and infers that it is of type float64. It also knows its length.

```python
@jit(nopython=True)
def foo(n):
    l = [i/2 for i in range(n)]
    return len(l)
```

Perhaps I am missing something, but it seems that numba could use the type and length of the list to create a tuple say if I called `tuple(l)` in the above function?

**Comment by ehsantn** [comment]
A list's length is not fixed but a tuple's length is fixed. The value of `n` is not constant in your example function.

**Comment by krrk** [comment]
I see. Well as you said earlier there are some other cases where I think you could infer the type of the tuple. Here are some of my ideas:

```python
@jit(nopython=True)
def foo(x):
    tuple([i for i in range(10)]) # most likely to be used
    tuple([i*x for i in range(10)]) # The type of x can be specified in the signature
    tuple(zip((True,)*10, (False,)*10))

    # The examples below could easily be created directly as tuples
    tuple(['c','h','a','r','s'])
    tuple([True,False,True])
```

**Comment by ehsantn** [comment]
Yes, these cases could be inferred. The zip one is probably a bit too complex for now but others are doable with reasonable effort.

**Comment by saulshanabrook** [comment]
Even supporting just the first case (`tuple(i for i in range(10)`) would be very helpful for me. [Right now](https://github.com/Quansight/numba-xnd/blob/ec5712819080204fe8352592b3a4acfeb3400103/numba_xnd/gumath.py#L73-L117) I have hard coded in functiond dispatcher up to 4 arguments, but if we had tuple generation, then I could create the arguments as a tuple and pass them in as `*args`.

**Comment by saulshanabrook** [comment]
Since we can already concatenate tuples and create them from literals, I wrote a function that creates a tuple creator dispatcher, given you know the length of the tuple at compile time:

```python
def create_tuple_creator(f, n):
    assert n > 0
    f = njit(f)
    @njit
    def creator(args):
        return (f(0, *args),)
    for i in range(1, n):
        # need to pass in creator and i to lambda to capture in scope
        creator = njit(lambda args, creator=creator, i=i: creator(args) + (f(i, *args),))
    return njit(lambda *args: creator(args))
```

Here are some examples of it solving some of the use cases in @krrk's post above:

```python
range_10 = create_tuple_creator(lambda i: i, 10)
range_10_x = create_tuple_creator(lambda i, x: i*x, 10)
true_10 = create_tuple_creator(lambda _: True, 10)
false_10 = create_tuple_creator(lambda _: False, 10)
zip_10 = create_tuple_creator(lambda i, l, r: (l[i], r[i]), 10)

@njit
def foo(x):
    print(range_10()) # tuple([i for i in range(10)])
    print(range_10_x(x)) # tuple([i*x for i in range(10)])
    print(zip_10(true_10(), false_10())),  # tuple(zip((True,)*10, (False,)*10))

    
foo(2)
```

**Comment by saulshanabrook** [comment]
Thinking about it, you could probably simplify usage a bit with `generated_jit` so you can pass in a constant integer inline for length instead of having to define that outside function.

**Comment by ehsantn** [comment]
I think this can be done similar to `np.array([...])` inlining [here](https://github.com/numba/numba/blob/master/numba/inline_closurecall.py#L522).

**Comment by cristipurdel** [comment]
Any chance to get an example with `tuple([iterable])` from documentation 3.5.2 [](http://numba.pydata.org/numba-doc/latest/cuda/cudapysupported.html#) built-in-functions isnide a cuda kernel?

**Comment by krrk** [comment]
@timothydmorton It is easy to work around this limitation by creating the arrays you need outside the jit'd function and pass them in as arguments.

**Comment by Julia1211** [comment]
> @jit(nopython=True)
> def g():
>     np.zeros([1,2,3])

> give the following error
> 
> ```python
> numba.errors.TypingError: Failed at nopython (nopython frontend)
> Invalid usage of Function(<built-in function zeros>) with parameters (list(int64))
> ```
> 
> This is perhaps not a big deal, but as a new user of numba and a fundamentally python user (I'm not used to using other languages) it is quite jarring to run into many situations where the pythonic way of programming has failed when attempting to use numba in nopython mode to gain the best speedup.
> 
I met this question now,and I can't understand how I can modify my code.Can you help me?

**Comment by Julia1211** [comment]
@krrk 
@jit(nopython=True)
def g():
np.zeros([1,2,3])

give the following error

numba.errors.TypingError: Failed at nopython (nopython frontend)
Invalid usage of Function(<built-in function zeros>) with parameters (list(int64))
This is perhaps not a big deal, but as a new user of numba and a fundamentally python user (I'm not used to using other languages) it is quite jarring to run into many situations where the pythonic way of programming has failed when attempting to use numba in nopython mode to gain the best speedup.

I met this question now,and I can't understand how I can modify my code.Can you help me?

**Comment by krrk** [comment]
@Julia1211 

In your example if you change the list to a tuple it will work:

```python
@jit(nopython=True)
def g():
    np.zeros((1,2,3))
```

If you want the shape of the array determined based on the input parameters that is possible as well:

```python
@jit(nopython=True)
def g(nx, ny):
    np.zeros((nx, ny))
```

however if you want to write a function that works on an arbitrary dimension array then you could pass the array into the jitted function:

```python
@jit(nopython=True)
def f(arr):
    arr += 1

arr = np.zeros([nx, ny, nz])
f(arr)
```

If your function is simple like the one I defined you could use `numba.vectorize` as well:

```python
@vectorize
def f(a):
    return a + 1

arr = np.zeros([nx, ny, nz])
new_arr = f(arr)
```

**Comment by Julia1211** [comment]
@krrk 
Thanks for your reply.And I have one other question that is there is a function A,and A will call an other function B during running program. np.ones([w,h]) is used in function B,so it has an error like that :

numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'B': cannot determine Numba type of <class 'function'>

B is so long that can't be merged into function A,so how can I change the code in this question?

**Comment by krrk** [comment]
@Julia1211 It looks like you're trying to call a regular python function
from a nopython jitted function which can only call other nopython jitted
functions. Please decorate your function `B` with `@jit(nopython=True)` as
well.

**Comment by krrk** [comment]
@Julia1211 I don't think your problem is related to this issue any longer. It seems you're trying to assign a new value to `roi` and that fails.

**Comment by senderle** [comment]
@krrk, I think you might have misunderstood @timothydmorton's comment. If you need to use a tuple to index a multidimensional array inside a loop, you're out of luck. There was an question on Stack Overflow about related issues, and I wrote up [some code with workarounds](https://stackoverflow.com/a/54635558/577088).

I'd be happy to know if there's a better way, but it does work.

**Comment by senderle** [comment]
I'd add that it seems particularly odd that you can't even create a tuple from a numpy array. It would seem a simple matter to infer the type of a tuple from the type of an array, unless the actual length of a tuple needs to be known at compile time. (But then... why would that be necessary for tuple a when it's not necessary for arrays?)

**Comment by seibert** [comment]
Tuples are currently represented as struct types in Numba, which is why type inference needs to know the length of the tuple at compile time (either because they are explicitly made in nopython mode, or because they are passed in as arguments to the nopython mode call).

**Comment by sjperkins** [comment]
I'm trying to adapt @saulshanabrook 's tuple creator:

> ```python
> def create_tuple_creator(f, n):
>     assert n > 0
>     f = njit(f)
>     @njit
>     def creator(args):
>         return (f(0, *args),)
>     for i in range(1, n):
>         # need to pass in creator and i to lambda to capture in scope
>         @njit
>         def creator(args, creator=creator, i=i):
>             return creator(args) + (f(i, *args),)
>     return njit(lambda *args: creator(args))
> ```

to produce variably shaped output tuples based on the presence of input parameters:


```python {.line-numbers}
import numpy as np
import numba
from numba import types, njit

def is_numba_type_none(arg):
    return (isinstance(arg, types.misc.NoneType) or
            (isinstance(arg, types.misc.Omitted) and arg.value is None))

def tuple_factory_factory(f, present):
    jitter = njit(nopython=True, nogil=True, cache=True)

    f = jitter(f)

    if present[0]:
        @jitter
        def creator(args):
            return (f(0, *args), )
    else:
        @jitter
        def creator(args):
            return ()

    for i in range(1, len(present)):
        if present[i]:
            @jitter
            def creator(args, creator=creator, i=i):
                return creator(args) + (f(i, *args),)

        else:
            @jitter
            def creator(args, creator=creator, i=i):
                return creator(args)

    return jitter(lambda *args: creator(args))

@numba.generated_jit(nopython=True, nogil=True, cache=True)
def fn(a=None, b=None):
    have_a = not is_numba_type_none(a)
    have_b = not is_numba_type_none(b)
    present = (have_a, have_b)
    output_tupler = tuple_factory_factory(lambda i, a: a[i], present)
    return lambda a=None, b=None: output_tupler((a, b))

a = np.arange(10)
b = np.arange(20, dtype=np.float32)

print(fn(a, b))
print(fn(None, b))
```
but compilation seems to fail on indexing a heterogenous tuple in `lambda i, a: a[i]`:

```
Invalid use of Function(<built-in function getitem>) with argument(s) of type(s):
((array(int64, 1d, C), array(float32, 1d, C)), Literal[int](0))
```

Any thoughts for getting over this hump? It seems possible to index heterogenous tuples in standard jitted code...

**Comment by jungerm2** [comment]
I know this is an old issue, but I simply wanted to point out that since dynamic tuple creation is not supported in nopython mode, it limits what functions we can use. For instance, in the following example I'm trying to implement a numba compatible `np.squeeze` method:

```python
@nb.jit
def squeeze(arr, axis=None):
    dims = arr.shape
    axis = axis if axis is not None else range(arr.ndim)
    dims = [d for i, d in enumerate(dims) if d != 1 and i in axis]
    return np.reshape(arr, dims)
```  

This fails because the only signature available for reshape is one where the shape argument is an int or _tuple_ of ints. Is there a workaround I'm not thinking of? Here clearly the size of the shape parameter `dims` is not known, but its size is upper-bounded easily as it will never be longer than `arr.shape`. Can this not be done statically at compile time?

**Comment by ziofil** [comment]
Is there a workaround if the length of a tuple is known and fixed?

**Comment by ulupo** [comment]
@ziofil I might have found an "unsafe" solution in the case in which the iterable to convert is an ndarray.  The idea being to use` to_fixed_tuple` which can be found in `numba.np.unsafe.ndarray`.  Although it is unsafe, it seems to work.  Note that in `to_fixed_tuple(arr, length)`, length must be a constant.  If it is not, but there is only a small set of possibilities for what this length might be, I found a solution which seems to work for me.  It uses `functools.lru_cache` to cache one (compiled) version of your njitted function makig use of `to_fixed_tuple`, per possible tuple length.  As an example, here is a way to convert a 2D ndarray `X` to a dictionary where each key is the tuple version of a row of `X` (and values are all NaN):

```
from functools import lru_cache

import numpy as np
from numba import njit
from numba.np.unsafe.ndarray import to_fixed_tuple

@lru_cache
def make_to_dict(length):

    @njit
    def to_dict(X):
        d = {}
        for row in X:
            d[to_fixed_tuple(row, length)] = np.nan

        return d

    return to_dict
```

You can try it with e.g.
```
to_dict = make_to_dict(2)
X = np.arange(10).reshape((5, 2))

d = to_dict(X)
```

On my machine at least, `to_dict` will only njit the first time around and a compiled cached version will be used from then on.

Wondering what the experts think!

**Comment by ogauthe** [comment]
> I am also finding myself needing to convert a 1d array into a tuple, in order to index a multidimensional array (need to pass a tuple of indices). Any update on this?

I have exactly the same problem as @timothydmorton, I can produce a list / array of indices, but I need a tuple to index a multidimensional array. Arrays are generated inside a loop, it is not possible for me to use them as arguments.

I used strides on `array.ravel()` as suggested by @senderle, however `ravel` may induce a copy if the array is not C-contiguous. I would really enjoy a simpler solution with native tuples.

**Comment by shaunc** [comment]
I'm hitting a related issue: https://github.com/numba/numba/issues/8248

I know very well what type of tuple I want, but am having trouble specifying it. Is there a "hidden constructor" somewhere that will build a struct that will come out to the correct tuple?
